pkg/circuit_breaker/circuit_breaker_collector.go

```diff
@@ -0,0 +1,158 @@
+package circuit_breaker
+
+import (
+	"context"
+	"github.com/frain-dev/convoy/config"
+	"github.com/frain-dev/convoy/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"time"
+)
+
+const namespace = "circuit_breaker"
+
+var (
+	cachedMetrics *Metrics
+	metricsConfig *config.MetricsConfiguration
+	lastRun       = time.Now()
+
+	circuitBreakerState = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "state"),
+		"The current state of the circuit breaker (0: Closed, 1: Half-Open, 2: Open)",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerRequests = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "requests_total"),
+		"Total number of requests processed by the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerFailures = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "failures_total"),
+		"Total number of failed requests processed by the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerSuccesses = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "successes_total"),
+		"Total number of successful requests processed by the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerFailureRate = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "failure_rate"),
+		"Current failure rate of the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerSuccessRate = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "success_rate"),
+		"Current success rate of the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerConsecutiveFailures = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "consecutive_failures"),
+		"Number of consecutive failures for the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerNotificationsSent = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "notifications_sent"),
+		"Number of notifications sent by the circuit breaker",
+		[]string{"key"}, nil,
```

Comment:
Is `key` the only required filter for now?

Reply:
Yeah, though, I think at least add the project_id for tenant segmentaton

Reply:
Ok, are we adding later or now?

Reply:
I'll do it now, I'll update it in the next commit